libraries/Microsoft.Bot.Builder.Dialogs/SkillDialog.cs

```diff
@@ -114,22 +133,51 @@ private static SkillDialogArgs ValidateBeginDialogArgs(object options)
             return dialogArgs;
         }
 
-        private async Task SendToSkillAsync(ITurnContext context, Activity activity, CancellationToken cancellationToken)
+        private async Task<Activity> SendToSkillAsync(ITurnContext context, Activity activity, CancellationToken cancellationToken)
         {
             // Create a conversationId to interact with the skill and send the activity
-            var skillConversationId = await DialogOptions.ConversationIdFactory.CreateSkillConversationIdAsync(activity.GetConversationReference(), cancellationToken).ConfigureAwait(false);
+            var conversationIdFactoryOptions = new SkillConversationIdFactoryOptions
+            {
+                FromBotOAuthScope = context.TurnState.Get<string>(BotAdapter.OAuthScopeKey),
+                FromBotId = DialogOptions.BotId,
+                Activity = activity,
+                BotFrameworkSkill = DialogOptions.Skill
+            };
+            var skillConversationId = await DialogOptions.ConversationIdFactory.CreateSkillConversationIdAsync(conversationIdFactoryOptions, cancellationToken).ConfigureAwait(false);
 
             // Always save state before forwarding
             // (the dialog stack won't get updated with the skillDialog and things won't work if you don't)
             var skillInfo = DialogOptions.Skill;
             await DialogOptions.ConversationState.SaveChangesAsync(context, true, cancellationToken).ConfigureAwait(false);
-            var response = await DialogOptions.SkillClient.PostActivityAsync(DialogOptions.BotId, skillInfo.AppId, skillInfo.SkillEndpoint, DialogOptions.SkillHostEndpoint, skillConversationId, activity, cancellationToken).ConfigureAwait(false);
+
+            var response = await DialogOptions.SkillClient.PostActivityAsync<Activity[]>(DialogOptions.BotId, skillInfo.AppId, skillInfo.SkillEndpoint, DialogOptions.SkillHostEndpoint, skillConversationId, activity, cancellationToken).ConfigureAwait(false);
 
             // Inspect the skill response status
             if (!(response.Status >= 200 && response.Status <= 299))
             {
                 throw new HttpRequestException($"Error invoking the skill id: \"{skillInfo.Id}\" at \"{skillInfo.SkillEndpoint}\" (status is {response.Status}). \r\n {response.Body}");
             }
+
+            Activity eocActivity = null;
+            if (activity.DeliveryMode == DeliveryModes.BufferedReplies && response.Body.Any())
+            {
+                // Process replies in the response.Body.
+                foreach (var fromSkillActivity in response.Body)
```

Comment:
nit: couple of minor things:
- technically the eoc _must_ be the last activity in the list right? (therefore you could just check the last element of the collection)
- and you could perhaps make use of the Send (array) overload on the turnContext. for example if the collection didn't finish with the eoc you just send the lot.

anyhow not clearer that these make for clearer code - so as I say just a nit.